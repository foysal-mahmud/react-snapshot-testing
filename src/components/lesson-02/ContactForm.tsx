import React, { useState } from 'react';

interface ContactFormProps {
  onSubmit?: (data: FormData) => Promise<void>;
  initialData?: Partial<FormData>;
}

interface FormData {
  name: string;
  email: string;
  subject: string;
  message: string;
  newsletter: boolean;
}

interface FormErrors {
  name?: string;
  email?: string;
  subject?: string;
  message?: string;
}

const ContactForm: React.FC<ContactFormProps> = ({ onSubmit, initialData }) => {
  const [formData, setFormData] = useState<FormData>({
    name: initialData?.name || '',
    email: initialData?.email || '',
    subject: initialData?.subject || '',
    message: initialData?.message || '',
    newsletter: initialData?.newsletter || false
  });

  const [errors, setErrors] = useState<FormErrors>({});
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [submitStatus, setSubmitStatus] = useState<'idle' | 'success' | 'error'>('idle');

  const validateForm = (): boolean => {
    const newErrors: FormErrors = {};

    if (!formData.name.trim()) {
      newErrors.name = 'Name is required';
    } else if (formData.name.trim().length < 2) {
      newErrors.name = 'Name must be at least 2 characters';
    }

    if (!formData.email.trim()) {
      newErrors.email = 'Email is required';
    } else if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(formData.email)) {
      newErrors.email = 'Please enter a valid email address';
    }

    if (!formData.subject.trim()) {
      newErrors.subject = 'Subject is required';
    }

    if (!formData.message.trim()) {
      newErrors.message = 'Message is required';
    } else if (formData.message.trim().length < 10) {
      newErrors.message = 'Message must be at least 10 characters';
    }

    setErrors(newErrors);
    return Object.keys(newErrors).length === 0;
  };

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => {
    const { name, value, type } = e.target;
    const checked = (e.target as HTMLInputElement).checked;

    setFormData(prev => ({
      ...prev,
      [name]: type === 'checkbox' ? checked : value
    }));

    // Clear error when user starts typing
    if (errors[name as keyof FormErrors]) {
      setErrors(prev => ({
        ...prev,
        [name]: undefined
      }));
    }
  };

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    
    if (!validateForm()) {
      return;
    }

    setIsSubmitting(true);
    setSubmitStatus('idle');

    try {
      if (onSubmit) {
        await onSubmit(formData);
      } else {
        // Simulate API call
        await new Promise(resolve => setTimeout(resolve, 1000));
      }
      
      setSubmitStatus('success');
      
      // Reset form on success
      setFormData({
        name: '',
        email: '',
        subject: '',
        message: '',
        newsletter: false
      });
    } catch (error) {
      setSubmitStatus('error');
    } finally {
      setIsSubmitting(false);
    }
  };

  const handleReset = () => {
    setFormData({
      name: '',
      email: '',
      subject: '',
      message: '',
      newsletter: false
    });
    setErrors({});
    setSubmitStatus('idle');
  };

  return (
    <form 
      onSubmit={handleSubmit} 
      className="contact-form"
      data-testid="contact-form"
      aria-label="Contact form"
    >
      <h2>Contact Us</h2>
      
      {submitStatus === 'success' && (
        <div 
          className="alert alert-success" 
          role="alert"
          data-testid="success-message"
        >
          Thank you! Your message has been sent successfully.
        </div>
      )}
      
      {submitStatus === 'error' && (
        <div 
          className="alert alert-error" 
          role="alert"
          data-testid="error-message"
        >
          Sorry, there was an error sending your message. Please try again.
        </div>
      )}

      <div className="form-group">
        <label htmlFor="name">Name *</label>
        <input
          type="text"
          id="name"
          name="name"
          value={formData.name}
          onChange={handleInputChange}
          className={errors.name ? 'error' : ''}
          aria-describedby={errors.name ? 'name-error' : undefined}
          disabled={isSubmitting}
        />
        {errors.name && (
          <span 
            id="name-error" 
            className="error-message" 
            role="alert"
            data-testid="name-error"
          >
            {errors.name}
          </span>
        )}
      </div>

      <div className="form-group">
        <label htmlFor="email">Email *</label>
        <input
          type="email"
          id="email"
          name="email"
          value={formData.email}
          onChange={handleInputChange}
          className={errors.email ? 'error' : ''}
          aria-describedby={errors.email ? 'email-error' : undefined}
          disabled={isSubmitting}
        />
        {errors.email && (
          <span 
            id="email-error" 
            className="error-message" 
            role="alert"
            data-testid="email-error"
          >
            {errors.email}
          </span>
        )}
      </div>

      <div className="form-group">
        <label htmlFor="subject">Subject *</label>
        <select
          id="subject"
          name="subject"
          value={formData.subject}
          onChange={handleInputChange}
          className={errors.subject ? 'error' : ''}
          aria-describedby={errors.subject ? 'subject-error' : undefined}
          disabled={isSubmitting}
        >
          <option value="">Please select a subject</option>
          <option value="general">General Inquiry</option>
          <option value="support">Technical Support</option>
          <option value="billing">Billing Question</option>
          <option value="feedback">Feedback</option>
        </select>
        {errors.subject && (
          <span 
            id="subject-error" 
            className="error-message" 
            role="alert"
            data-testid="subject-error"
          >
            {errors.subject}
          </span>
        )}
      </div>

      <div className="form-group">
        <label htmlFor="message">Message *</label>
        <textarea
          id="message"
          name="message"
          value={formData.message}
          onChange={handleInputChange}
          rows={5}
          className={errors.message ? 'error' : ''}
          aria-describedby={errors.message ? 'message-error' : undefined}
          disabled={isSubmitting}
        />
        {errors.message && (
          <span 
            id="message-error" 
            className="error-message" 
            role="alert"
            data-testid="message-error"
          >
            {errors.message}
          </span>
        )}
      </div>

      <div className="form-group">
        <label className="checkbox-label">
          <input
            type="checkbox"
            name="newsletter"
            checked={formData.newsletter}
            onChange={handleInputChange}
            disabled={isSubmitting}
          />
          Subscribe to our newsletter
        </label>
      </div>

      <div className="form-actions">
        <button
          type="submit"
          disabled={isSubmitting}
          className="btn btn-primary"
          aria-describedby="submit-help"
        >
          {isSubmitting ? 'Sending...' : 'Send Message'}
        </button>
        
        <button
          type="button"
          onClick={handleReset}
          disabled={isSubmitting}
          className="btn btn-secondary"
        >
          Reset
        </button>
      </div>
      
      <p id="submit-help" className="help-text">
        * Required fields
      </p>
    </form>
  );
};

export default ContactForm; 