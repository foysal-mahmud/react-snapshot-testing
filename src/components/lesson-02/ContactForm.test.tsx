import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ContactForm from './ContactForm';

describe('Lesson 2: ContactForm Component - Forms & Async Testing', () => {
  
  // 📝 FORM TESTING FUNDAMENTALS
  describe('Form Elements and Labels', () => {
    test('finds form elements by label text (best practice)', () => {
      render(<ContactForm />);
      
      // This is the GOLD STANDARD for form testing
      expect(screen.getByLabelText(/name/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/subject/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/message/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/subscribe to our newsletter/i)).toBeInTheDocument();
    });

    test('finds form by role and aria-label', () => {
      render(<ContactForm />);
      
      const form = screen.getByRole('form', { name: /contact form/i });
      expect(form).toBeInTheDocument();
    });

    test('finds submit button by role', () => {
      render(<ContactForm />);
      
      const submitButton = screen.getByRole('button', { name: /send message/i });
      const resetButton = screen.getByRole('button', { name: /reset/i });
      
      expect(submitButton).toBeInTheDocument();
      expect(resetButton).toBeInTheDocument();
      expect(submitButton).toHaveAttribute('type', 'submit');
    });

    test('finds select dropdown by role', () => {
      render(<ContactForm />);
      
      const subjectSelect = screen.getByRole('combobox', { name: /subject/i });
      expect(subjectSelect).toBeInTheDocument();
    });
  });

  // 👆 USER INTERACTIONS WITH FORMS
  describe('User Interactions with userEvent', () => {
    test('allows user to type in text inputs', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      const nameInput = screen.getByLabelText(/name/i);
      const emailInput = screen.getByLabelText(/email/i);
      
      await user.type(nameInput, 'John Doe');
      await user.type(emailInput, 'john@example.com');
      
      expect(nameInput).toHaveValue('John Doe');
      expect(emailInput).toHaveValue('john@example.com');
    });

    test('allows user to select from dropdown', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      const subjectSelect = screen.getByLabelText(/subject/i);
      
      await user.selectOptions(subjectSelect, 'support');
      expect(subjectSelect).toHaveValue('support');
      
      // Test that the correct option is selected
      expect(screen.getByRole('option', { name: /technical support/i })).toBeInTheDocument();
    });

    test('allows user to type in textarea', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      const messageTextarea = screen.getByLabelText(/message/i);
      const testMessage = 'This is a test message for the contact form.';
      
      await user.type(messageTextarea, testMessage);
      expect(messageTextarea).toHaveValue(testMessage);
    });

    test('allows user to toggle checkbox', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      const newsletterCheckbox = screen.getByLabelText(/subscribe to our newsletter/i);
      
      expect(newsletterCheckbox).not.toBeChecked();
      
      await user.click(newsletterCheckbox);
      expect(newsletterCheckbox).toBeChecked();
      
      await user.click(newsletterCheckbox);
      expect(newsletterCheckbox).not.toBeChecked();
    });
  });

  // ✅ FORM VALIDATION TESTING
  describe('Form Validation', () => {
    test('shows validation errors for empty required fields', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      const submitButton = screen.getByRole('button', { name: /send message/i });
      
      // Submit empty form
      await user.click(submitButton);
      
      // Check for validation errors using different query methods
      expect(screen.getByTestId('name-error')).toHaveTextContent('Name is required');
      expect(screen.getByTestId('email-error')).toHaveTextContent('Email is required');
      expect(screen.getByTestId('subject-error')).toHaveTextContent('Subject is required');
      expect(screen.getByTestId('message-error')).toHaveTextContent('Message is required');
    });

    test('shows specific validation errors for invalid input', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      const nameInput = screen.getByLabelText(/name/i);
      const emailInput = screen.getByLabelText(/email/i);
      const messageInput = screen.getByLabelText(/message/i);
      const submitButton = screen.getByRole('button', { name: /send message/i });
      
      // Enter invalid data
      await user.type(nameInput, 'A'); // Too short
      await user.type(emailInput, 'invalid-email'); // Invalid format
      await user.type(messageInput, 'Short'); // Too short
      
      await user.click(submitButton);
      
      expect(screen.getByTestId('name-error')).toHaveTextContent('Name must be at least 2 characters');
      expect(screen.getByTestId('email-error')).toHaveTextContent('Please enter a valid email address');
      expect(screen.getByTestId('message-error')).toHaveTextContent('Message must be at least 10 characters');
    });

    test('clears errors when user starts typing', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      const nameInput = screen.getByLabelText(/name/i);
      const submitButton = screen.getByRole('button', { name: /send message/i });
      
      // Trigger validation error
      await user.click(submitButton);
      expect(screen.getByTestId('name-error')).toBeInTheDocument();
      
      // Start typing to clear error
      await user.type(nameInput, 'J');
      expect(screen.queryByTestId('name-error')).not.toBeInTheDocument();
    });
  });

  // ⏰ ASYNC BEHAVIOR TESTING
  describe('Async Form Submission', () => {
    test('shows loading state during form submission', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      // Fill out valid form
      await user.type(screen.getByLabelText(/name/i), 'John Doe');
      await user.type(screen.getByLabelText(/email/i), 'john@example.com');
      await user.selectOptions(screen.getByLabelText(/subject/i), 'general');
      await user.type(screen.getByLabelText(/message/i), 'This is a test message for the form.');
      
      const submitButton = screen.getByRole('button', { name: /send message/i });
      
      // Submit form
      await user.click(submitButton);
      
      // Check loading state immediately
      expect(screen.getByRole('button', { name: /sending/i })).toBeInTheDocument();
      expect(submitButton).toBeDisabled();
      
      // Wait for submission to complete
      await waitFor(() => {
        expect(screen.getByTestId('success-message')).toBeInTheDocument();
      }, { timeout: 2000 });
    });

    test('shows success message after successful submission', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      // Fill out valid form
      await user.type(screen.getByLabelText(/name/i), 'Jane Smith');
      await user.type(screen.getByLabelText(/email/i), 'jane@example.com');
      await user.selectOptions(screen.getByLabelText(/subject/i), 'feedback');
      await user.type(screen.getByLabelText(/message/i), 'Great service! Keep up the good work.');
      
      await user.click(screen.getByRole('button', { name: /send message/i }));
      
      // Wait for success message using findBy (for async elements)
      const successMessage = await screen.findByTestId('success-message');
      expect(successMessage).toHaveTextContent('Thank you! Your message has been sent successfully.');
    });

    test('resets form after successful submission', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      const nameInput = screen.getByLabelText(/name/i);
      const emailInput = screen.getByLabelText(/email/i);
      const messageInput = screen.getByLabelText(/message/i);
      
      // Fill and submit form
      await user.type(nameInput, 'Test User');
      await user.type(emailInput, 'test@example.com');
      await user.selectOptions(screen.getByLabelText(/subject/i), 'general');
      await user.type(messageInput, 'Test message content here.');
      
      await user.click(screen.getByRole('button', { name: /send message/i }));
      
      // Wait for success and check form is reset
      await screen.findByTestId('success-message');
      
      expect(nameInput).toHaveValue('');
      expect(emailInput).toHaveValue('');
      expect(messageInput).toHaveValue('');
    });

    test('handles submission errors', async () => {
      const mockOnSubmit = jest.fn().mockRejectedValue(new Error('Network error'));
      const user = userEvent.setup();
      
      render(<ContactForm onSubmit={mockOnSubmit} />);
      
      // Fill out valid form
      await user.type(screen.getByLabelText(/name/i), 'Error Test');
      await user.type(screen.getByLabelText(/email/i), 'error@example.com');
      await user.selectOptions(screen.getByLabelText(/subject/i), 'support');
      await user.type(screen.getByLabelText(/message/i), 'This should trigger an error.');
      
      await user.click(screen.getByRole('button', { name: /send message/i }));
      
      // Wait for error message
      const errorMessage = await screen.findByTestId('error-message');
      expect(errorMessage).toHaveTextContent('Sorry, there was an error sending your message. Please try again.');
    });
  });

  // 🔄 RESET FUNCTIONALITY
  describe('Reset Functionality', () => {
    test('resets form when reset button is clicked', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      // Fill out form
      await user.type(screen.getByLabelText(/name/i), 'Reset Test');
      await user.type(screen.getByLabelText(/email/i), 'reset@example.com');
      await user.click(screen.getByLabelText(/subscribe to our newsletter/i));
      
      // Verify form has data
      expect(screen.getByLabelText(/name/i)).toHaveValue('Reset Test');
      expect(screen.getByLabelText(/subscribe to our newsletter/i)).toBeChecked();
      
      // Click reset
      await user.click(screen.getByRole('button', { name: /reset/i }));
      
      // Verify form is reset
      expect(screen.getByLabelText(/name/i)).toHaveValue('');
      expect(screen.getByLabelText(/email/i)).toHaveValue('');
      expect(screen.getByLabelText(/subscribe to our newsletter/i)).not.toBeChecked();
    });
  });

  // 📸 SNAPSHOT TESTS - Different Form States
  describe('Lesson 2: Form State Snapshots', () => {
    test('captures empty form state', () => {
      render(<ContactForm />);
      const form = screen.getByTestId('contact-form');
      expect(form).toMatchSnapshot();
    });

    test('captures form with validation errors', async () => {
      const user = userEvent.setup();
      render(<ContactForm />);
      
      // Trigger validation errors
      await user.click(screen.getByRole('button', { name: /send message/i }));
      
      const form = screen.getByTestId('contact-form');
      expect(form).toMatchSnapshot();
    });

    test('captures form with initial data', () => {
      const initialData = {
        name: 'John Doe',
        email: 'john@example.com',
        subject: 'general',
        message: 'Pre-filled message',
        newsletter: true
      };
      
      render(<ContactForm initialData={initialData} />);
      const form = screen.getByTestId('contact-form');
      expect(form).toMatchSnapshot();
    });
  });
}); 