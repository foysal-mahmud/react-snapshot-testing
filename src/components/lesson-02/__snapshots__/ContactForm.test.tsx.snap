// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson 2: ContactForm Component - Forms & Async Testing Lesson 2: Form State Snapshots captures empty form state 1`] = `
<form
  aria-label="Contact form"
  class="contact-form"
  data-testid="contact-form"
>
  <h2>
    Contact Us
  </h2>
  <div
    class="form-group"
  >
    <label
      for="name"
    >
      Name *
    </label>
    <input
      class=""
      id="name"
      name="name"
      type="text"
      value=""
    />
  </div>
  <div
    class="form-group"
  >
    <label
      for="email"
    >
      Email *
    </label>
    <input
      class=""
      id="email"
      name="email"
      type="email"
      value=""
    />
  </div>
  <div
    class="form-group"
  >
    <label
      for="subject"
    >
      Subject *
    </label>
    <select
      class=""
      id="subject"
      name="subject"
    >
      <option
        value=""
      >
        Please select a subject
      </option>
      <option
        value="general"
      >
        General Inquiry
      </option>
      <option
        value="support"
      >
        Technical Support
      </option>
      <option
        value="billing"
      >
        Billing Question
      </option>
      <option
        value="feedback"
      >
        Feedback
      </option>
    </select>
  </div>
  <div
    class="form-group"
  >
    <label
      for="message"
    >
      Message *
    </label>
    <textarea
      class=""
      id="message"
      name="message"
      rows="5"
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="checkbox-label"
    >
      <input
        name="newsletter"
        type="checkbox"
      />
      Subscribe to our newsletter
    </label>
  </div>
  <div
    class="form-actions"
  >
    <button
      aria-describedby="submit-help"
      class="btn btn-primary"
      type="submit"
    >
      Send Message
    </button>
    <button
      class="btn btn-secondary"
      type="button"
    >
      Reset
    </button>
  </div>
  <p
    class="help-text"
    id="submit-help"
  >
    * Required fields
  </p>
</form>
`;

exports[`Lesson 2: ContactForm Component - Forms & Async Testing Lesson 2: Form State Snapshots captures form with initial data 1`] = `
<form
  aria-label="Contact form"
  class="contact-form"
  data-testid="contact-form"
>
  <h2>
    Contact Us
  </h2>
  <div
    class="form-group"
  >
    <label
      for="name"
    >
      Name *
    </label>
    <input
      class=""
      id="name"
      name="name"
      type="text"
      value="John Doe"
    />
  </div>
  <div
    class="form-group"
  >
    <label
      for="email"
    >
      Email *
    </label>
    <input
      class=""
      id="email"
      name="email"
      type="email"
      value="john@example.com"
    />
  </div>
  <div
    class="form-group"
  >
    <label
      for="subject"
    >
      Subject *
    </label>
    <select
      class=""
      id="subject"
      name="subject"
    >
      <option
        value=""
      >
        Please select a subject
      </option>
      <option
        value="general"
      >
        General Inquiry
      </option>
      <option
        value="support"
      >
        Technical Support
      </option>
      <option
        value="billing"
      >
        Billing Question
      </option>
      <option
        value="feedback"
      >
        Feedback
      </option>
    </select>
  </div>
  <div
    class="form-group"
  >
    <label
      for="message"
    >
      Message *
    </label>
    <textarea
      class=""
      id="message"
      name="message"
      rows="5"
    >
      Pre-filled message
    </textarea>
  </div>
  <div
    class="form-group"
  >
    <label
      class="checkbox-label"
    >
      <input
        checked=""
        name="newsletter"
        type="checkbox"
      />
      Subscribe to our newsletter
    </label>
  </div>
  <div
    class="form-actions"
  >
    <button
      aria-describedby="submit-help"
      class="btn btn-primary"
      type="submit"
    >
      Send Message
    </button>
    <button
      class="btn btn-secondary"
      type="button"
    >
      Reset
    </button>
  </div>
  <p
    class="help-text"
    id="submit-help"
  >
    * Required fields
  </p>
</form>
`;

exports[`Lesson 2: ContactForm Component - Forms & Async Testing Lesson 2: Form State Snapshots captures form with validation errors 1`] = `
<form
  aria-label="Contact form"
  class="contact-form"
  data-testid="contact-form"
>
  <h2>
    Contact Us
  </h2>
  <div
    class="form-group"
  >
    <label
      for="name"
    >
      Name *
    </label>
    <input
      aria-describedby="name-error"
      class="error"
      id="name"
      name="name"
      type="text"
      value=""
    />
    <span
      class="error-message"
      data-testid="name-error"
      id="name-error"
      role="alert"
    >
      Name is required
    </span>
  </div>
  <div
    class="form-group"
  >
    <label
      for="email"
    >
      Email *
    </label>
    <input
      aria-describedby="email-error"
      class="error"
      id="email"
      name="email"
      type="email"
      value=""
    />
    <span
      class="error-message"
      data-testid="email-error"
      id="email-error"
      role="alert"
    >
      Email is required
    </span>
  </div>
  <div
    class="form-group"
  >
    <label
      for="subject"
    >
      Subject *
    </label>
    <select
      aria-describedby="subject-error"
      class="error"
      id="subject"
      name="subject"
    >
      <option
        value=""
      >
        Please select a subject
      </option>
      <option
        value="general"
      >
        General Inquiry
      </option>
      <option
        value="support"
      >
        Technical Support
      </option>
      <option
        value="billing"
      >
        Billing Question
      </option>
      <option
        value="feedback"
      >
        Feedback
      </option>
    </select>
    <span
      class="error-message"
      data-testid="subject-error"
      id="subject-error"
      role="alert"
    >
      Subject is required
    </span>
  </div>
  <div
    class="form-group"
  >
    <label
      for="message"
    >
      Message *
    </label>
    <textarea
      aria-describedby="message-error"
      class="error"
      id="message"
      name="message"
      rows="5"
    />
    <span
      class="error-message"
      data-testid="message-error"
      id="message-error"
      role="alert"
    >
      Message is required
    </span>
  </div>
  <div
    class="form-group"
  >
    <label
      class="checkbox-label"
    >
      <input
        name="newsletter"
        type="checkbox"
      />
      Subscribe to our newsletter
    </label>
  </div>
  <div
    class="form-actions"
  >
    <button
      aria-describedby="submit-help"
      class="btn btn-primary"
      type="submit"
    >
      Send Message
    </button>
    <button
      class="btn btn-secondary"
      type="button"
    >
      Reset
    </button>
  </div>
  <p
    class="help-text"
    id="submit-help"
  >
    * Required fields
  </p>
</form>
`;
